### PR TITLE
handle unicode/blobdb issue

### PR DESCRIPTION
**Pull request by sravfeyn**
https://sentry.io/dimagi/commcarehq/issues/774216550/

@mkangia

### CODE REVIEW

File: custom/icds_reports/reports/disha.py [hunk 1]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
...
+            if written_count != total:
+                file_obj.write(",")
+            logger.info("Processed {count}/{total} batches".format(count=count, total=num_batches))
+        file_obj.write("]}")
 
     def build_export_json(self):
         with TransientTempfile() as temp_path:
-            self._write_data_in_chunks(temp_path)
-            with open(temp_path, 'r', encoding='utf-8') as f:
+            with open(temp_path, 'w+b') as f:
```
Comment (mkangia) [inline custom/icds_reports/reports/disha.py]: does `wb` work too?

File: custom/icds_reports/tests/test_disha_dump.py [hunk 2]
```diff
@@ -20,14 +20,15 @@ class CountableList(list):
...
         disha_get_rows_mock.return_value = CountableList(data)
 
         month = date(2018, 8, 1)
         state = 'Andhra pradesh'
         with TransientTempfile() as temp_path:
             dump = DishaDump(state, month)
-            dump._write_data_in_chunks(temp_path)
+            with open(temp_path, 'w+b') as f:
+                dump._write_data_in_chunks(f)
             with open(temp_path, 'r', encoding='utf-8') as f:
```
Comment (mkangia) [inline custom/icds_reports/tests/test_disha_dump.py]: just out of curiosity,
you write in bytes but read unicodes and are them comparing the data with non unicode `"d\xc3\xa9f"`, that works?

File: custom/icds_reports/reports/disha.py [hunk 3]
```diff
@@ -29,7 +29,7 @@ def __init__(self, state_name, month):
         self.month = month
 
     def _blob_id(self):
-        return 'disha_dump-{}-{}.json'.format(self.state_name.replace(" ", ""), self.month.strftime('%Y-%m-%d'))
+        return 'disha_dump-{}-{}.json'.format(self.state_name.encode('ascii', 'ignore').replace(" ", ""), self.month.strftime('%Y-%m-%d'))
```
Comment (mkangia) [inline custom/icds_reports/reports/disha.py]: What was the state name that caused this? `Madhya Pradesh`?

**Review by mkangia — APPROVED** [review]

File: custom/icds_reports/reports/disha.py (hunk 3 above)
Comment (sravfeyn) [reply custom/icds_reports/reports/disha.py]: There was no issue here, I just took a precaution. This would break if the state-name as unicode only chars.

File: custom/icds_reports/reports/disha.py (hunk 1 above)
Comment (sravfeyn) [reply custom/icds_reports/reports/disha.py]: Yeah.

File: custom/icds_reports/tests/test_disha_dump.py (hunk 2 above)
Comment (sravfeyn) [reply custom/icds_reports/tests/test_disha_dump.py]: The data is written is utf8 encoded bytes, so when we read back, we specify the same encoding to read back those bytes, so it works.

Comment (mkangia) [reply custom/icds_reports/tests/test_disha_dump.py]: so f.read() gives you back 'utf8 encoded' data but we compare it with "d\xc3\xa9f" and that comes to be equal?

Comment (sravfeyn) [reply custom/icds_reports/tests/test_disha_dump.py]: Yeah. Because `d\xc3\xa9f` is also utf8 encoded string.

File: custom/icds_reports/reports/disha.py (hunk 3 above)
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: This will not work on Python 3. `.encode(...)` always returns bytes, but `.replace(" ", "")` is a text (Unicode) operation, so it will fail on Python 3. Should be
```py
return 'disha_dump-{}-{}.json'.format(
    self.state_name.encode('ascii', 'ignore').decode('ascii').replace(" ", ""),
    self.month.strftime('%Y-%m-%d'),
)
```
`encode` removes non-ASCII characters while converting to bytes and `decode` converts bytes to text.

File: custom/icds_reports/reports/disha.py [hunk 4]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
         #   }
         columns = self._get_columns()
         indicators = self._get_rows()
-        metadata_line = '{{'\
+        metadata_line = u'{{'\
```
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: This is not necessary because `from __future__ import unicode_literals` is in this module.

File: custom/icds_reports/reports/disha.py [hunk 5]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
...
+        num_batches = 10
+        for count, (_, end, total, chunk) in enumerate(batch_qs(indicators, num_batches=num_batches)):
+            chunk_string = json.dumps(list(chunk), ensure_ascii=False).encode('utf8')
+            # chunk is list of lists, so skip enclosing brackets
+            file_obj.write(chunk_string[1:-1])
+            written_count += len(chunk)
+            if written_count != total:
+                file_obj.write(",")
+            logger.info("Processed {count}/{total} batches".format(count=count, total=num_batches))
+        file_obj.write("]}")
```
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: should be `file_obj.write(b"]}")`

File: custom/icds_reports/reports/disha.py [hunk 6]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
...
-                if written_count != total:
-                    f.write(",")
-                logger.info("Processed {count}/{total} batches".format(count=count, total=num_batches))
-            f.write("]}")
+                columns=json.dumps(columns, ensure_ascii=False)).encode('utf8')
+        file_obj.write(metadata_line)
+        written_count = 0
+        num_batches = 10
+        for count, (_, end, total, chunk) in enumerate(batch_qs(indicators, num_batches=num_batches)):
+            chunk_string = json.dumps(list(chunk), ensure_ascii=False).encode('utf8')
```
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: The return type of `json.dumps(...)` is different on Python 2 and 3 (due to a bug in Python 2). This code will work on both Python 2 and 3, but it's kinda sneaky to do an unnecessary encode on Python 2 like that since the result is already bytes (see also [this comment](https://github.com/dimagi/langcodes/pull/19#discussion_r189239788)).

My recommendation here would be to revert the changes in this file and make a simpler change here:
```py
chunk_string = json.dumps(list(chunk), ensure_ascii=False)
if six.PY2:
    chunk_string = chunk_string.decode('utf-8')
```
With that change all the code should just work as it was written before, using `open(temp_path, 'w', encoding='utf-8')`. The `six.PY2` check + decode could eventually be removed when we switch to Python 3 (although it will not hurt to let it there).

Alternately, if you want to keep the code as you have it (with some fixes I'm suggesting in other comments), you could do this instead:
```py
chunk_string = json.dumps(list(chunk), ensure_ascii=False)
if six.PY3:
    chunk_string = chunk_string.encode('utf-8')
```

File: custom/icds_reports/reports/disha.py [hunk 7]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
...
+            written_count += len(chunk)
+            if written_count != total:
+                file_obj.write(",")
+            logger.info("Processed {count}/{total} batches".format(count=count, total=num_batches))
+        file_obj.write("]}")
 
     def build_export_json(self):
         with TransientTempfile() as temp_path:
-            self._write_data_in_chunks(temp_path)
-            with open(temp_path, 'r', encoding='utf-8') as f:
```
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: This was definitely wrong. Up to you if you want to keep the change of reading/writing from the same file object, but you could revert the changes here and simply change this to
```py
with open(temp_path, 'rb') as f:
```

File: custom/icds_reports/reports/disha.py [hunk 8]
```diff
@@ -81,32 +81,32 @@ def _write_data_in_chunks(self, temp_path):
...
+        file_obj.write(metadata_line)
+        written_count = 0
+        num_batches = 10
+        for count, (_, end, total, chunk) in enumerate(batch_qs(indicators, num_batches=num_batches)):
+            chunk_string = json.dumps(list(chunk), ensure_ascii=False).encode('utf8')
+            # chunk is list of lists, so skip enclosing brackets
+            file_obj.write(chunk_string[1:-1])
+            written_count += len(chunk)
+            if written_count != total:
+                file_obj.write(",")
```
Comment (millerdev) [inline custom/icds_reports/reports/disha.py]: should be `file_obj.write(b",")`

File: custom/icds_reports/tests/test_disha_dump.py [hunk 9]
```diff
@@ -20,14 +20,15 @@ class CountableList(list):
             def count(self, *args, **kwargs):
                 return len(self)
 
-        data = [['a'], ['b'], ['b\u0105z']]
+        data = [['a'], ['b'], ["d\xc3\xa9f"]]
```
Comment (millerdev) [inline custom/icds_reports/tests/test_disha_dump.py]: Why was this change necessary?

File: custom/icds_reports/reports/disha.py (hunk 3 above)
Comment (sravfeyn) [reply custom/icds_reports/reports/disha.py]: Ha, thanks for this note! I have fixed this in a different way https://github.com/dimagi/commcare-hq/pull/22513/files#diff-48e45a26c4763ddeafc50cc7f7e97a12R34
